### PR TITLE
Feature/soft actor critic

### PR DESCRIPTION
**Pull request by jdrotleff**
#### Summary of additions and changes

* Add a Soft Actor-Critic implementation for continuous-action off-policy RL.
  * New `SACAgent` with replay-buffer based updates.
  * New `SoftActorCriticLoss` with actor, twin-critic, entropy-temperature, and Polyak target updates.
  * New SAC TD target helper via `TDReturnsSAC`.

* Add replay-buffer infrastructure for off-policy training.
  * New `Transition` dataclass.
  * New fixed-size `ReplayBuffer`.
  * New transition trajectory storage for optional HDF5 debugging/export.

* Add flexible multi-network support for actor/critic/temperature-style algorithms.
  * New `MultiFlaxModel` container for multiple Flax modules and train states.
  * Support for target parameters used by SAC critics.

* Update continuous Gaussian sampling for SAC-style policies.
  * Supports Gaussian actor logits split into mean and log-std.
  * Supports tanh correction for log-probabilities.
  * Supports deterministic deployment mode.

* Add `NoExplorationPolicy` for stochastic-policy algorithms where exploration is intrinsic to the policy.

* Add `UniversalTrainer`.
  * Supports continuous and episodic training through one trainer.
  * Uses duck typing to distinguish off-policy agents from on-policy agents.
  * Updates off-policy agents step-wise and on-policy agents episode-wise.
  * Preserves existing Espresso trainer behavior through compatibility tests.

* Extend test coverage.
  * Unit tests for SAC loss, SAC agent, replay buffer, Gaussian sampling, transition storage, SAC TD returns, `MultiFlaxModel`, and `UniversalTrainer`.
  * Integration smoke test for SAC replay-buffer to JIT-loss dataflow.
  * Espresso compatibility test for `UniversalTrainer`.

#### Reviewer notes / design choices

* SAC uses a `MultiFlaxModel` container instead of extending `FlaxModel`, because SAC needs independently optimized actor, critic, target critic, and entropy-temperature state.

* The SAC loss intentionally freezes critic parameters during the actor-loss Q-value pass. Gradients should flow through the sampled action into the actor, but not update critic weights from actor loss. This is covered by dedicated tests.

* `UniversalTrainer` distinguishes off-policy agents by duck typing (`replay_buffer is not None`) instead of adding a new base class. This keeps compatibility with existing agents but is a design point worth reviewing. Additionally to that, we can discuss if we still need the `EpisodicTrainer` and `ContinuousTrainer`, if we have an universal one.

* `NoExplorationPolicy` is added for stochastic-policy algorithms where exploration is already handled by the policy. SAC does not currently require it directly, but it is useful for `ActionSelector`/future continuous stochastic-policy setups where adding external noise might make action log-probabilities inconsistent.

* The neural-network API is more flexible than before: SAC passes observations as keyword arguments and passes `actions` explicitly to the critic. This should make it applicable to a broad set of architectures, including GNN-style models, as long as their `apply` signatures follow this contract.

#### How to test this pull request
Run the CI. Locally, the main focused check is:

```
pytest \
  CI/unit_tests/agents/test_sac_agent.py \
  CI/unit_tests/losses/test_sac_loss.py \
  CI/unit_tests/sampling_strategies/test_gaussian.py \
  CI/unit_tests/replay_buffer/test_replay_buffer.py \
  CI/unit_tests/trainers/test_universal_trainer.py \
  CI/integration_tests/test_sac_pipeline_integration.py \
  -q
```

### CODE REVIEW

**Comment by jdrotleff** [comment]
Last refactor changes the SAC agent to use a single user-provided `FlaxModel` module instead
of the multi-network wrapper. Removed it completely.

The provided network now requires explicit `actor(...)`, `critic(...)`, and `alpha()`
methods, because training needs separate entry points for action
sampling, Q-value evaluation, and temperature updates.

File: swarmrl/sampling_strategies/gaussian_distribution.py [hunk 1]
```diff
@@ -81,50 +119,57 @@ def __call__(
...
             log_probs = None
         else:
-            if subkey is None:
-                subkey = rng_key
-            if subkey is None:
-                subkey = jax.random.PRNGKey(onp.random.randint(0, 1236534623))
+            if rng_key is None:
+                raise ValueError("A valid JAX PRNGKey is required during training.")
+
             log_std = jnp.clip(
```
[the remark this answers is not in the document]
Comment (jdrotleff) [reply swarmrl/sampling_strategies/gaussian_distribution.py]: resolved by adding soft clip
